src/pybind/mgr/dashboard/frontend/src/app/shared/directives/form-loading.directive.spec.ts

```diff
@@ -0,0 +1,85 @@
+import { CdForm } from '../forms/cd-form';
+import { SharedModule } from '../shared.module';
+import { FormLoadingDirective } from './form-loading.directive';
+
+@Component({ selector: 'cd-test-cmp', template: '<span *cdFormLoading="loading">foo</span>' })
+class TestComponent extends CdForm {
+  constructor() {
+    super();
+  }
+}
+
+describe('FormLoadingDirective', () => {
+  let component: TestComponent;
+  let fixture: ComponentFixture<any>;
+
+  configureTestBed({
+    declarations: [TestComponent],
+    imports: [AlertModule.forRoot(), SharedModule],
+    providers: [i18nProviders]
+  });
+
+  afterEach(() => {
+    fixture = null;
+  });
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(TestComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
+
+  it('should create an instance', () => {
+    const directive = new FormLoadingDirective(null, null, null, null);
+    expect(directive).toBeTruthy();
+  });
+
+  it('should show loading component by default', () => {
+    expect(fixture.debugElement.queryAll(By.css('span')).length).toEqual(0);
+    expect(fixture.debugElement.queryAll(By.css('cd-alert-panel')).length).toEqual(0);
+    expect(fixture.debugElement.queryAll(By.css('cd-loading-panel')).length).toEqual(1);
+
+    const alert = fixture.debugElement.nativeElement.querySelector('cd-loading-panel alert');
+    expect(alert.textContent).toBe('Loading form data...');
+  });
+
+  it('should show error component when calling loadingError()', () => {
+    component.loadingError();
+    fixture.detectChanges();
+
+    expect(fixture.debugElement.queryAll(By.css('span')).length).toEqual(0);
+    expect(fixture.debugElement.queryAll(By.css('cd-alert-panel')).length).toEqual(1);
+    expect(fixture.debugElement.queryAll(By.css('cd-loading-panel')).length).toEqual(0);
+
+    const alert = fixture.debugElement.nativeElement.querySelector('cd-alert-panel .alert-panel-text');
+    expect(alert.textContent).toBe('Form data could not be loaded.');
+  });
+
+  it('should show original component when calling loadingReady()', () => {
+    component.loadingReady();
+    fixture.detectChanges();
+
+    expect(fixture.debugElement.queryAll(By.css('span')).length).toEqual(1);
+    expect(fixture.debugElement.queryAll(By.css('cd-alert-panel')).length).toEqual(0);
+    expect(fixture.debugElement.queryAll(By.css('cd-loading-panel')).length).toEqual(0);
+
+    const alert = fixture.debugElement.nativeElement.querySelector('span');
+    expect(alert.textContent).toBe('foo');
+  });
+
+  it('should show nothing when calling loadingNone()', () => {
+    component.loadingNone();
+    fixture.detectChanges();
+
+    expect(fixture.debugElement.queryAll(By.css('span')).length).toEqual(0);
+    expect(fixture.debugElement.queryAll(By.css('cd-alert-panel')).length).toEqual(0);
+    expect(fixture.debugElement.queryAll(By.css('cd-loading-panel')).length).toEqual(0);
```

Comment:
Looks good but good you please move the redundant part into a function like `expectShown(component, error, loading)`?

Reply:
Done